src/text.c

```diff
@@ -0,0 +1,358 @@
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '=':
+				SKIP;
+				fwrite ("\\=", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+
+		fwrite ("=", 1, 1, f);
+
+		// write and escape value
+		p = v;
+		n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+		fwrite ("\n", 1, 1, f);
+	}
+	ls_free (l);
+
+	// sub-namespaces
+	l = s->ns;
+	if (sort) {
+		l = ls_clone (l);
+		ls_sort (l, cmp_ns);
+	}
+	SdbNs *ns;
+	ls_foreach (l, it, ns) {
+		fwrite ("\n", 1, 1, f);
+		ls_push (path, ns->name);
+		text_fsave (ns->sdb, f, sort, path);
+		ls_pop (path);
+	}
+	if (l != s->ns) {
+		ls_free (l);
+	}
+
+#undef FLUSH
+	return true;
+}
+
+SDB_API bool sdb_text_fsave(Sdb *s, FILE *f, bool sort) {
+	SdbList *path = ls_new ();
+	if (!path) {
+		return false;
+	}
+	bool r = text_fsave (s, f, sort, path);
+	ls_free (path);
+	return r;
+}
+
+SDB_API bool sdb_text_save(Sdb *s, const char *file, bool sort) {
+	FILE *f = fopen (file, "w");
+	if (!f) {
+		return false;
+	}
+	bool r = sdb_text_fsave (s, f, sort);
+	fclose (f);
+	return r;
+}
+
+SDB_API bool sdb_text_fload(Sdb *s, FILE *f) {
+	size_t bufsz = 16; // TODO: more
+	char *buf = malloc (bufsz);
+	if (!buf) {
+		return false;
+	}
+	bool ret = true;
+	Sdb *cur_db = s;
+	// can't use pointers here because buf may be moved on realloc
+	size_t pos = 0; // current processing position in the buffer
+	size_t token_begin = 0; // beginning of the currently processed token in the buffer
+	size_t shift = 0; // amount to shift chars to the left (from unescaping)
+	SdbList/*<size_t>*/ *path = ls_new ();
+	if (!path) {
+		free (buf);
+		return false;
+	}
+	enum { STATE_NEWLINE, STATE_PATH, STATE_KEY, STATE_VALUE } state = STATE_NEWLINE;
+	bool unescape = false; // whether the prev char was a backslash, i.e. the current one is escaped
+	while (true) {
+		// realloc if no space
+		if (pos >= bufsz - 1) {
+			size_t newsz = bufsz + bufsz;
+			if (newsz < bufsz) {
+				ret = false;
+				break;
+			}
+			bufsz = newsz;
+			char *newbuf = realloc (buf, bufsz);
+			if (!newbuf) {
+				ret = false;
+				break;
+			}
+			buf = newbuf;
+		}
+		// read full buffer (leave 1 for null-terminating)
+		size_t read = fread (buf + pos, 1, bufsz - 1 - pos, f);
+		if (!read && !feof (f)) {
+			// error
+			ret = false;
+			break;
+		}
+		// process buffer
+		size_t buf_filled = pos + read;
+		if (!read) {
+			goto flush;
+		}
+		while (pos < buf_filled) {
+			if (state == STATE_NEWLINE) {
+				// at the start of a line, decide whether it's a path or a k=v
+				// by whether there is a leading slash.
+				if (buf[pos] == '/') {
+					state = STATE_PATH;
+					token_begin = 1;
+					pos++;
+					continue;
+				} else {
+					state = STATE_KEY;
+				}
+			}
+			if (buf[pos] == '\n') {
```

Comment:
mmm i was expecting a sort of `switch` here, not a "random" check against `\n`. I think it would be much easier to follow if you use a switch on the state and then check the characters in the buffer according to the state you are in.

Reply:
switch on the state would need code duplication, but I think after splitting it into multiple functions, it is better anyway.